File: nvidia/c-cpp/cublas/examples/openacc_cublas.cpp
#include <cublas_v2.h>
#include <cuda_runtime.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <vector>

int main() {
  const int m = 1024, n = 1024, k = 1024;
  const double alpha = 1.0, beta = 1.0;
  std::vector<double> a(static_cast<std::size_t>(m) * k, 1.0);
  std::vector<double> b(static_cast<std::size_t>(k) * n, 1.0);
  std::vector<double> c(static_cast<std::size_t>(m) * n, 1.0);
  double *ap = a.data(), *bp = b.data(), *cp = c.data();
  const std::size_t ac = a.size(), bc = b.size(), cc = c.size();
  cublasHandle_t handle = nullptr;
  if (cublasCreate(&handle) != CUBLAS_STATUS_SUCCESS) {
    std::fprintf(stderr, "cublasCreate failed\n");
    return EXIT_FAILURE;
  }
  bool ok = true;
#pragma acc data copyin(ap[0 : ac], bp[0 : bc]) copy(cp[0 : cc])
  {
#pragma acc host_data use_device(ap, bp, cp)
    {
      ok = cublasDgemm(handle, CUBLAS_OP_N, CUBLAS_OP_N, m, n, k, &alpha, ap, m,
                       bp, k, &beta, cp, m) == CUBLAS_STATUS_SUCCESS;
    }
    ok = ok && cudaDeviceSynchronize() == cudaSuccess;
  }
  ok = ok && cublasDestroy(handle) == CUBLAS_STATUS_SUCCESS;
  double max_error = 0.0;
  for (double value : c)
    max_error = std::max(max_error, std::fabs(value - (k + 1.0)));
  std::printf("OpenACC-managed cuBLAS DGEMM complete; max error = %.17g\n",
              max_error);
  return ok && max_error <= 1.0e-10 ? EXIT_SUCCESS : EXIT_FAILURE;
}

File: nvidia/fortran/cublas/examples/openacc_cublas.f90
program openacc_cublas
    use cudafor
    use cublas_v2
    use, intrinsic :: ieee_arithmetic, only: ieee_is_finite
    implicit none
    integer :: allocation_status

    integer, parameter :: m = 1024
    integer, parameter :: n = 1024
    integer, parameter :: k = 1024
    real(8) :: alpha, beta
    integer :: istat
    type(cublasHandle) :: handle
    real(8), allocatable :: mat_A(:,:), mat_B(:,:), mat_C(:,:)
    alpha = 1.0d0; beta  = 1.0d0

    allocate(mat_A(m,k), mat_B(k,n), mat_C(m,n), stat=allocation_status)

    if (allocation_status /= 0) error stop 'allocation failed'
    mat_A = 1.0d0; mat_B = 1.0d0; mat_C = 1.0d0

    istat = cublasCreate(handle)
    if (istat /= 0) error stop 'cublasCreate failed'

    !$acc data copyin(mat_A(1:m,1:k), mat_B(1:k,1:n)) &
    !$acc     copy(mat_C(1:m,1:n))
        !$acc host_data use_device(mat_A, mat_B, mat_C)
            istat = cublasDgemm(handle, &
                CUBLAS_OP_N, CUBLAS_OP_N, m, n, k, &
                alpha, mat_A, m, mat_B, k, &
                beta,  mat_C, m)
    if (istat /= 0) error stop 'cublasDgemm failed'
        !$acc end host_data
        istat = cudaDeviceSynchronize()
    if (istat /= 0) error stop 'cudaDeviceSynchronize failed'
    !$acc end data
    print *, "mat_C(1,1) = ", mat_C(1,1)
    if (.not. all(ieee_is_finite(mat_C))) error stop 'nonfinite DGEMM result'
    if (maxval(abs(mat_C - (alpha*real(k,8)+beta))) > 1.0d-10*real(k,8)) &
        error stop 'DGEMM verification failed'
    print *, 'verification PASS'
    istat = cublasDestroy(handle)
    if (istat /= 0) error stop 'cublasDestroy failed'
    deallocate(mat_A, mat_B, mat_C)
end program openacc_cublas
